presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ActualProperties.java

```diff
@@ -153,17 +154,21 @@ public boolean isStreamRepartitionEffective(Collection<Symbol> keys)
 
     public ActualProperties translate(Function<Symbol, Optional<Symbol>> translator)
     {
-        Map<Symbol, NullableValue> translatedConstants = new HashMap<>();
-        for (Map.Entry<Symbol, NullableValue> entry : constants.entrySet()) {
-            Optional<Symbol> translatedKey = translator.apply(entry.getKey());
-            if (translatedKey.isPresent()) {
-                translatedConstants.put(translatedKey.get(), entry.getValue());
-            }
-        }
         return builder()
-                .global(global.translate(translator, symbol -> Optional.ofNullable(constants.get(symbol))))
+                .global(global.translate(new Partitioning.Translator(translator, symbol -> Optional.ofNullable(constants.get(symbol)), expression -> Optional.empty())))
                 .local(LocalProperties.translate(localProperties, translator))
-                .constants(translatedConstants)
+                .constants(translateConstants(translator))
+                .build();
+    }
+
+    public ActualProperties translate(
+            Function<Symbol, Optional<Symbol>> translator,
+            Function<Expression, Optional<Symbol>> expressionTranslator)
```

Comment:
I come up with the similar thought with https://github.com/prestodb/presto/pull/12013#discussion_r241567120 . I didn't see follow-ups on that thread, so I assume this is addressed offline?


I can see why we want to have a separate `expressionTranslator` though. Translator over symbol is much lightweight and easy to reason, so it might worthy to have e dedicated symbol translator :) . Just want to see if this is how previous discussion addressed?

Reply:
To the comment you referenced. We introduced a `Translator` class to wrap 3 different translators. The problem of introducing a forth one: Symbol -> Expression translator, is that now Symbol can potentially have two translators, so which one to use? You'd actually only either use one or the other, never both. I can't think of any tidy way of implementing that. Since FULL_JOIN is the only place that would use a symbol -> Expression translator, I think writing it the current way is a ok trade off for now.

Reply:
@rongrong : Sorry for the confusion :). I am not suggesting to do anything different. But just curious about  if there are any follow-ups on this previous comment thread offline, and curious about what alternatives are discussed.